Updated logic to support duplicate note

bubudrc opened the pull request:
### Fix
This PR will add the feature that let users to duplicate an existing note.  Will fix: [issue-1066](https://github.com/Automattic/simplenote-macos/issues/1066)
The idea was to create a new note, and set the content and tags of the existing one.

For that, I must to add new menu items and updated the layout and conditions to not duplicate code

### Test
> 1. Select any note of the list
> 2. Press right button
> 3. Select "Duplicate Note" option

> 1. Select any note of the list
> 2. In the Editor view, select the "more options" menu item
> 3. Select "Duplicate Note" option

> 1. Select any note of the list
> 2. In main menu select Note
> 3. Select "Duplicate Note" option

![duplicate_note](https://user-images.githubusercontent.com/619159/226000243-8b52d9f4-6871-4089-bf21-54a56db880bd.gif)


### Review
> Only one developer are required to review these changes, but anyone can perform the review.

### Release
> `RELEASE-NOTES.txt` was updated with:
> 
> > Added Duplicate Note support


### Notes
Because this is a new complete feature, and I'm not part of the company, there are some steps that need your interaction (if you want to merge this). 
The new "Duplicate Note" needs localization strings.
The new Duplicate Note, needs update the analytics with a new event, to track this and know from where was called (Notes list, Main menu, Editor Menu)

bubudrc committed:
Updated the layout to add a new item for the different menues. Updated the logic to support duplicate note

bubudrc committed:
Fixed minor issue with action link

jleandroperez commented on Simplenote/NoteEditorViewController+Swift.swift:
```diff
@@ -417,6 +423,12 @@ extension NoteEditorViewController: NSMenuItemValidation {
     }
 
+    func validateListDuplicateNoteMenuItem(_ item: NSMenuItem) -> Bool {
+        // TODO: Add localization keys+values for this new item
```
Would it be possible to remove this TODO? We've got tooling that takes care of that 😄 

Thank you!

jleandroperez replied:
```suggestion
```

jleandroperez commented on Simplenote/NoteEditorViewController.m:
```diff
@@ -599,4 +581,50 @@ - (void)insertChecklistAction:(id)sender
 
+#pragma mark - New Note
+
+- (void)duplicateNote:(nullable Note *)sender
```
Question:  `duplicateNote` seems to be always invoked with `nil` as a parameter.

WDYT if we remove the parameter altogether?

bubudrc replied:
duplicateNote is called with the selected note on NoteViewController.swift:856
```
@IBAction
    func duplicateNote(_ sender: Any) {
        guard let note = selectedNotes.first else {
            return
        }

        noteEditorViewController.duplicateNote(note)
        // TODO: Set correct analytic event
    }
```
We could remove the parameter and always use the `self.note` value of NoteEditorViewController. It's that ok?

jleandroperez replied:
- That sounds super, we could just drop the parameter 😄 
- Perhaps `validateListDuplicateNoteMenuItem` should also check for `selectedNotes.count == 1`. Allowing for Duplicate, when there's more than one entry selected (and only duplicating the first one) would probably not be what I'd expect. WDYT?
- Do you think we could rename the `@IBAction func duplicateNote` to `duplicateNoteWasPressed` ? (so that it looks exactly the same as the other IBAction handlers).

Also, thank you sir!

bubudrc replied:
Done.
I'm changed the methods names and also added the `selectedNotes.count == 1` condition, not only on `validateListDuplicateNoteMenuItem` but also on `validateSystemDuplicateNoteMenuItem` to keep a consistence on the logic.

BTW: I renamed `validateListDuplicateNoteMenuItem` to `validateEditDuplicateNoteMenuItem` to match the existing method names

jleandroperez commented on Simplenote/NoteListViewController.swift:
```diff
@@ -731,6 +734,12 @@ extension NoteListViewController: NSMenuItemValidation {
     }
 
+    func validateListDuplicateNoteMenuItem(_ item: NSMenuItem) -> Bool {
+        // TODO: Add localization keys+values for this new item
```
```suggestion
```

jleandroperez commented on Simplenote/NoteEditorViewController.m:
```diff
@@ -599,4 +581,50 @@ - (void)insertChecklistAction:(id)sender
+    }
+}
+
+- (void) createNoteWasPressed:(nullable Note *)sender {
```
Since this API isn't wired as an IBAction (and `sender` is used usually as the parameter name for the component that's doing the invocation), and UI components won't ever invoke this directly, WDYT if this was renamed as...

- `createNoteFromNote:(nullable Note *)oldNote`
- `createNoteWithNote:(nullable Note *)oldNote`
- `createNewNoteWithContent:(nullable Note *)content`
- `buildNewNoteWithContent:(nullable Note *)content`
- [Another cooler alternative here]

Thank you!

bubudrc replied:
I like your alternatives. For sure I'm will change it :)

jleandroperez replied:
Thank you!

jleandroperez reviewed (changes requested):
Hi there @bubudrc! Sending few nipticky notes. Other than that, looks great!!

**Thanks in advance!!**

P.s.: I'll get to your second PR hopefully this week, apologies about the delay.

bubudrc commented:
Hi @jleandroperez . Thanks for taking the time to review the PR and sent me the notes.
I'm making the suggested changes, but I'm need a confirmation in one of that changes. With that answer, I'm will be able to update the PR.

jleandroperez commented:
Apologies about the slow review @bubudrc, and again, thank you for the PR!!

bubudrc committed:
Updated methods name
Removed no needed comments
Added validations to only duplicate notes were only one is selected
Fixed problems with keys

bubudrc commented:
One more thing. Now, we can duplicate empty notes, but that has none sense. I'm know that [there are an open issue](https://github.com/Automattic/simplenote-macos/issues/1064) asking to remove the empty notes, so fix that issue could be a good opportunity to improve the behavior and avoid duplicate empty notes :)

jleandroperez commented:
I do agree, perhaps allowing to duplicate an existing note feels odd (but yeah, I also agree, what I would expect is to simply not be able to have empty notes, otherwise the behavior would feel inconsistent!).

Thank you for helping us improve Simplenote @bubudrc !!

:shipit:
